### PR TITLE
Theme: Add a stats api endpoint.

### PR DESCRIPTION
**Pull request by StevenDufresne**
This PR adds API endpoints to retrieve information about themes used to inform theme review improvements. 

At this point in time, this endpoint is planned to be used in conjunction with [add pr to chart block] and therefore returns data in Google Charts format.

## Endpoints

Note: All data is grouped by month and assumed the theme was published.

| URL                           | Description |
|:----------------------------- |:----------- |
| /themes/v1/stats/byThemeType  | Used to compared FSE theme vs Classic theme uploads            |
| /themes/v1/stats/bySegment    | Groups themes into 3 categories at upload time: (1 theme, 2 - 4 themes, 5+ themes)            |
| /themes/v1/stats/byAuthorType | New vs Returning authors at upload time            |
| /themes/v1/stats/reviewDays   | How long it takes to review a theme on average            |

## Considerations

**Security**
- Endpoints permissions are set to `current_user_can( 'edit_posts' )`. Is that a good idea? Ultimately this can probably be public.

**Privacy**
- Theme author names are included in the results. This isn't private information, as you can see that information in various ways in the theme directory. Just wanted to mention it.

**Are there any other concerns with creating endpoints for this data?**

### CODE REVIEW

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+			)
+		);
+	}
+
+	/**
+	 * Returns a list of published themes.
+	 *
+	 * @return void
+	 */
+	public function get_uploaded_themes() {
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php:173]: Should this function have some kind of caching given everything is ultimately pulled from here? Even if it's only an hour.

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+			FROM {$wpdb->posts} p
+			JOIN {$wpdb->postmeta} pm ON p.ID = pm.post_id AND pm.meta_key = '_author'
+			LEFT JOIN {$wpdb->term_relationships} tr ON p.ID = tr.object_id
+			LEFT JOIN {$wpdb->term_taxonomy} tt ON tr.term_taxonomy_id = tt.term_taxonomy_id AND tt.taxonomy = 'post_tag'
+			LEFT JOIN {$wpdb->terms} t ON tt.term_id = t.term_id
+
+			WHERE p.post_type = 'repopackage' and p.post_status = 'publish'
+
+			Group by p.ID
+			Order By published_on asc
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]: nitpicking, it's best to keep SQLs as uppercase for SQL keywords & readability.
```suggestion
			WHERE p.post_type = 'repopackage' AND p.post_status = 'publish'

			GROUP BY p.ID
			ORDER BY published_on ASC
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+		global $wpdb;
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery
+		return $wpdb->get_results(
+			"	
+			SELECT p.post_name as postName,
+				p.post_author as post_author,
+				substring_index( substring_index( pm.meta_value, '\"', 4 ), '\"', -1 ) AS author,
+				LEFT(p.post_date, 7) as published_on,
+				group_concat( t.slug SEPARATOR ', ' ) as tags
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]: Same as below.
```suggestion
			SELECT p.post_name AS postName,
				p.post_author AS post_author,
				substring_index( substring_index( pm.meta_value, '\"', 4 ), '\"', -1 ) AS author,
				LEFT(p.post_date, 7) AS published_on,
				group_concat( t.slug SEPARATOR ', ' ) AS tags
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+					SELECT p.post_name,
+					pm.meta_value,
+					substring_index( substring_index( pm.meta_value, '\"', 4 ), '\"', -1 ) AS upload_date,
+					p.post_date_gmt AS published_date,
+					DATEDIFF( p.post_date_gmt, substring_index( substring_index( pm.meta_value, '\"', 4 ), '\"', -1 ) ) as days_to_review
+			
+                    FROM {$wpdb->posts} p
+                    JOIN {$wpdb->postmeta} pm ON p.ID = pm.post_id AND pm.meta_key = '_upload_date'
+			
+					WHERE p.post_type = 'repopackage' and p.post_status = 'publish'
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]: Just the whitespace
```suggestion
					DATEDIFF( p.post_date_gmt, substring_index( substring_index( pm.meta_value, '\"', 4 ), '\"', -1 ) ) as days_to_review
			
					FROM {$wpdb->posts} p
					JOIN {$wpdb->postmeta} pm ON p.ID = pm.post_id AND pm.meta_key = '_upload_date'
			
					WHERE p.post_type = 'repopackage' and p.post_status = 'publish'
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+	 * @return void
+	 */
+	public function get_average_review_days( $start_date ) {
+		global $wpdb;
+
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"	
+				SELECT LEFT( upload_date, 7 ) as ym, AVG(days_to_review) as average_days_to_review
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]:
```suggestion
				SELECT LEFT( upload_date, 7 ) AS ym, AVG(days_to_review) AS average_days_to_review
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+			}
+		);
+	}
+
+	/**
+	 * Returns a list averages review times by month.
+	 *
+	 * @return void
+	 */
+	public function get_average_review_days( $start_date ) {
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php:124]: Should this function have some kind of caching given everything is ultimately pulled from here? Even if it's only an hour.

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+	/**
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_by_theme_type( $request ) {
+		$start_date = $this->get_start_date( $request );
+		$rows       = $this->get_uploaded_themes( $start_date );
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]:
```suggestion
		$rows       = $this->get_uploaded_themes();
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+	/**
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_by_segment( $request ) {
+		$start_date = $this->get_start_date( $request );
+		$rows       = $this->get_uploaded_themes( $start_date );
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]:
```suggestion
		$rows       = $this->get_uploaded_themes();
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+
+	/**
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_by_author_type( $request ) {
+		$start_date = $this->get_start_date( $request );
+		$rows       = $this->get_uploaded_themes( $start_date );
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]:
```suggestion
		$rows       = $this->get_uploaded_themes();
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+
+	/**
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_items( $request ) {
+		$start_date = $this->get_start_date( $request );
+		$rows       = $this->get_uploaded_themes( $start_date );
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]:
```suggestion
		$rows       = $this->get_uploaded_themes();
```

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+	/**
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_by_segment( $request ) {
+		$start_date = $this->get_start_date( $request );
+		$rows       = $this->get_uploaded_themes( $start_date );
+		$authorMap  = array(); // Temporarily store themes per author
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]: Generally nice to keep all the variables in the same case type for readability, ie. `$author_map`

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+		}
+
+		$out = array();
+		foreach ( $map as $year_month => $value ) {
+			array_push( $out, array( $year_month, $value['full-site-editing'], $value['classic'] ) );
+		}
+
+		$out = $this->filter_items_before( $out, $start_date );
+
+		return new WP_REST_Response( array_values( $out ), \WP_Http::OK );
```
Comment (dd32) [inline wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]: You can just skip including the return code for the responses, I went looking for where a different return-code would be set... or just use `rest_ensure_response()` instead to handle it all magically.. 🤷 not important really this is fine.
```suggestion
		return rest_ensure_response( array_values( $out ) );
```

**Review by dd32 — APPROVED** [review]
Bunch of nitpicks, but no alarm bells anywhere.

The `edit_posts` cap is probably fine to start with, but is probably the wrong cap check long-term as you've noted.

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php (hunk 11 above)
Comment (StevenDufresne) [reply wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php]: Ah dang. yes.

File: wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php (hunk 1 above)
Comment (StevenDufresne) [reply wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php:173]: Maybe create a intermediary function that uses `wp_cache_get`?

Comment (dd32) [reply wordpress.org/public_html/wp-content/plugins/theme-directory/rest-api/class-theme-review-stats.php:173]: I would just throw a wp_cache_get/set before/after the SQL.

```suggestion
	public function get_uploaded_themes() {
		$cached = wp_cache_get( __METHOD__, 'some-group' );
		if ( $cached ) {
			return $cached;
		}
```

**Pull request closed without merging**